File: python/newsfetcher.py
import requests
import logging
from datetime import datetime, timedelta
import os
from dotenv import load_dotenv
from datetime import datetime, timezone


class CryptoNewsFetcher:
    """Handles fetching news and Twitter information for cryptocurrencies."""

    def __init__(self):
        # Load API keys from environment variables

        self.news_api_key = os.getenv("NEWS_API_KEY")


    def fetch_crypto_news(self, keyword, max_articles=5):
        """Fetches top news articles about a specific cryptocurrency using NewsAPI."""
        url = "https://newsapi.org/v2/everything"
        today = datetime.now(timezone.utc).strftime('%Y-%m-%d')
        yesterday = (datetime.now(timezone.utc) - timedelta(days=1)).strftime('%Y-%m-%d')

        params = {
            "q": keyword,
            "language": "fr",  # French articles
            "sortBy": "publishedAt",
            "from": yesterday,  # Articles from yesterday onward
            "apiKey": self.news_api_key,
            "pageSize": max_articles
        }
        try:
            response = requests.get(url, params=params)
            logging.info(f"NewsAPI Response... = {response.status_code == 200}");
            if response.status_code == 200:
                logging.info(f"NewsAPI for {keyword}");
                articles = response.json().get("articles", [])
                news = [{"title": article["title"], "url": article["url"]} for article in articles]
                return news
            else:
                logging.error(f"NewsAPI Error for {keyword}: {response.status_code} - {response.text}")
                return []
        except Exception as e:
            logging.error(f"Error fetching news for {keyword}: {e}")
            return []



if __name__ == "__main__":
    analyzer = CryptoNewsFetcher()
    analyzer.fetch_crypto_news(keyword="Bitcoin", max_articles=1)

File: newsFetcher.js
const axios = require('axios');
const { TwitterApi } = require('twitter-api-v2');

// NewsFetcher Class
class NewsFetcher {
    constructor(newsApiKey, twitterBearerToken) {
        this.newsApiKey = newsApiKey;
        this.twitterClient = new TwitterApi(twitterBearerToken);
    }

    // Fetch news articles
    async fetchNews(query, maxArticles = 2) {
        const url = "https://newsapi.org/v2/everything";
        const today = new Date().toISOString().split('T')[0];
        const yesterday = new Date(Date.now() - 86400000).toISOString().split('T')[0];

        try {
            const response = await axios.get(url, {
                params: {
                    q: query,
                    sortBy: "publishedAt",
                    from: yesterday,
                    to: today,
                    apiKey: this.newsApiKey,
                    pageSize: maxArticles,
                },
            });

            if (response.status === 200) {
                console.info(`[newsFetcher.js] Response.status === ${response.status}`)
                console.info('[newsFetcher.js] Full API Response:', response.data);
                return response.data.articles.map(article => ({
                    title: article.title,
                    url: article.url,
                }));
            } else {
                console.error(`[newsFetcher.js] NewsAPI error: ${response.status}`);
                return [];
            }
        } catch (error) {
            console.error('[newsFetcher.js] Error fetching news:', error);
            return [];
        }
    }

    // Fetch tweets
    async fetchTweets(query, maxTweets = 2) {
        try {
            const response = await this.twitterClient.v2.search(query, {
                "tweet.fields": "created_at",
                max_results: maxTweets,
            });
            return response.data || [];
        } catch (error) {
            console.error('Error fetching tweets:', error);
            return [];
        }
    }
}

module.exports = NewsFetcher;